persistence/javadsl/src/main/scala/com/lightbend/lagom/internal/javadsl/persistence/ReadSideImpl.scala

```diff
@@ -71,27 +75,116 @@ private[lagom] class ReadSideImpl @Inject() (
+          config.role,
+          config.minBackoff,
+          config.maxBackoff,
+          config.randomBackoffFactor
+        )
+
+      val readSideProps =
+        ReadSideTagHolderActor.props(
+          config,
+          globalPrepareTask,
+          processorFactory,
+          registry,
+          eventClass
+        )
 
       val shardingSettings = ClusterShardingSettings(system).withRole(config.role)
 
       ClusterDistribution(system).start(
         readSideName,
-        backoffProps,
+        readSideProps,
         entityIds.toSet,
         ClusterDistributionSettings(system).copy(clusterShardingSettings = shardingSettings)
       )
     }
 
   }
 }
+
+import akka.actor.{ Actor, Props }
+
+object ReadSideTagHolderActor {
+  def props[Event <: AggregateEvent[Event]](
+    config:            ReadSideConfig,
+    globalPrepareTask: ClusterStartupTask,
+    processorFactory:  () => ReadSideProcessor[Event],
+    registry:          PersistentEntityRegistry,
+    eventClass:        Class[Event]
+  )(implicit mat: Materializer): Props =
+    Props(
+      new ReadSideTagHolderActor(
+        config,
+        globalPrepareTask: ClusterStartupTask, processorFactory,
+        registry,
+        eventClass
+      )(
+        mat
+      )
+    )
+
+  // A ReadSideTagHolder is a 1:1 relation to a ReadSideActor. The difference is that
+  // the ReadSideActor may die and be respawned by a supervisor while the ReadSideTagHolderActor
+  // is a sharded actor that may only be respawned by EnsureActive.
+  // because those two respawn operations have different frequencies the TagHolder is provided
+  // as a cache for the ReadSideActor to retart faster when it crashes.
+  case object GetTag
+  case class CachedTag(tag: EntityId)
+
+}
+
+// A ReadSideActor consumes a shard of the whole event stream. That operation
+// happens inside a node of cluster. Akka Cluster Sharding is used to distribute
+// the tags across the cluster. The distribution of tags happens on a keep-alive
+// loop provided in ClusterDistribution.
+// So, each node start a ClusterDistribution and a ShardRegion. Then the
+// ClusterDistribution runs an infinite loop that constantly sends  all required
+// tags to the ShradRegion. The ShardRegion distributes the information to each
+// appropriate node where each local ShardRegion redirects each
+class ReadSideTagHolderActor[Event <: AggregateEvent[Event]](
+  config:            ReadSideConfig,
+  globalPrepareTask: ClusterStartupTask,
+  processorFactory:  () => ReadSideProcessor[Event],
+  registry:          PersistentEntityRegistry,
+  eventClass:        Class[Event]
+)(implicit mat: Materializer) extends Actor {
+  override def receive = {
+    case EnsureActive(tag) =>
+
+      val processorProps =
+        ReadSideActor.props(
+          processorFactory,
+          registry.eventStream[Event],
+          eventClass,
+          globalPrepareTask,
+          config.globalPrepareTimeout,
+          self
+        )
```

Comment:
If we pass the tag to the props of `ReadSideActor` it won't need to ask it to the tag holder. 
Instead of `ReadSideTagHolderActor` we could have a `ReadSideShard`.

`ReadSideShard` is the real sharded actor, but it does nothing except the initialisation of the `BackoffSupervisor` wrapping `ReadSideActor`.

Reply:
Maybe the actor receiving the `EnsureActive` message should be the one responsible for running the `globalPrepare`. Once this is done, it initialises the underlying actors.

As such, the `ReadSideActor` is responsible for running the `prepare` and start the streaming.

Reply:
Did we give up on using `RestartSource`?
The `prepare` method is supposed to be called more than once as discussed [here](https://github.com/lagom/lagom/issues/999#issuecomment-331144806).

Maybe we can simply remove `ReadSideTagHolderActor` and the `BackoffSupervisor`. 
Once  we ensure the actor is alive and pass it the tag it must query, the stream will be recovered on failure and we won't be dependent on `EnsureActive` to restart the stream.

Reply:
There's still an option for `RestartXxx` see my comment in https://github.com/lagom/lagom/issues/999#issuecomment-331418173

Reply:
If we moved the `globalPrepare` to the outer actor, then the outer actor needs to be wrapped in a BackoffSupervisor.

How about:

 - an outer actor that builds the inner actor with the tag in the props 
 - the inner actor is wrapped in a BackoffSupervisor to ensure `globalPrepare` completes successfully.
 - the outer actor handles the `EnsureActive` message blocking it from propagating downstream when the inner actor is already built.
 - the inner actor uses `RestartXxxx` so once built the stream self-heals 
 - the inner actor's stream invokes `prepare` in the closures (twice?)

Reply:
Hmmm, the `ClusterStartupTask` already wraps the `task` in a `BackoffSupervisor`.